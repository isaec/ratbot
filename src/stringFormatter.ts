export const formatList = (
  list: { toString: () => string }[],
  joiner = ", ",
  finalJoiner = " and ",
  finalJoinerMinLength = 2
): string => {
  if (list.length === 0) return "";

  const stringList = list.map((x) => x.toString());

  if (stringList.length < finalJoinerMinLength || finalJoinerMinLength < 2)
    return stringList.join(joiner);

  return `${stringList
    .slice(0, -1)
    .join(joiner)}${finalJoiner}${stringList.slice(-1)}`;
};

export const formatRange = (numbers: number[]): string => {
  const sorted = numbers.sort((a, b) => a - b);
  const elements: Array<number | string> = [];
  let currentStart: number | undefined = undefined;
  for (let i = 0; i < sorted.length; i++) {
    const current = sorted[i];
    const next = sorted[i + 1];

    if (currentStart === undefined) currentStart = current;
    if (next === current + 1) continue; // continue if the next number is sequential

    if (currentStart === current) {
      elements.push(currentStart);
    } else {
      elements.push(`${currentStart}-${current}`);
    }
    currentStart = undefined;
  }
  return formatList(elements, ", ", " and ", 3);
};
