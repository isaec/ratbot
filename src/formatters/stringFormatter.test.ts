import { assert, describe, expect, it, test } from "vitest";
import { formatList, formatRange } from "./stringFormatter";

const numbers = (...arr: number[]) => ({ numbers: arr });

describe("formatRange", () => {
  test.each([
    numbers(1, 2, 3, 4, 5, 6, 7, 8, 9, 10),
    numbers(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11),
    numbers(1, 3, 4, 5, 6, 7, 8, 9, 10, 11),
    numbers(10, 24, 27, 14, 0, 8, 6, 20, 5, 2, 19, 15, 4, 28, 21, 30),
  ])(`formatRange(%o)`, (obj) => {
    expect(formatRange(obj.numbers)).toMatchSnapshot();
  });
  it.each([
    [numbers(1, 2, 3, 4, 5, 6, 7, 8, 9, 10), "1-10"],
    [numbers(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11), "1-11"],
    [numbers(1, 3, 4, 5, 6, 7, 8, 9, 10, 11), "1, 3-11"],
  ])("formats array of %o to %s", ({ numbers }, expected) => {
    expect(formatRange(numbers)).toBe(expected);
  });
});

describe("formatList", () => {
  it.each([
    [[1, 2, 3], "1, 2 and 3"],
    [[1, 2, 3, 4], "1, 2, 3 and 4"],
    [[1, 2, 3, 4, 5], "1, 2, 3, 4 and 5"],
    [[1, "3"], "1 and 3"],
  ])(`formatList(%o)`, (arr, expected) => {
    expect(formatList(arr)).toBe(expected);
  });
});
