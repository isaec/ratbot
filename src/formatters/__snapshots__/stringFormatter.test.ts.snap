// Vitest Snapshot v1

exports[`formatRange > formatRange({
  numbers: [
    10,           24,
    27,           14,
    0,            8,
    6,            20,
    5,            2,
    19,           15,
    4,            28,
    21,           30,
    [length]: 16
  ]
}) 1`] = `"0, 2, 4-6, 8, 10, 14-15, 19-21, 24, 27-28 and 30"`;

exports[`formatRange > formatRange({ numbers: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, [length]: 10 ] }) 1`] = `"1-10"`;

exports[`formatRange > formatRange({ numbers: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, [length]: 11 ] }) 1`] = `"1-11"`;

exports[`formatRange > formatRange({ numbers: [ 1, 3, 4, 5, 6, 7, 8, 9, 10, 11, [length]: 10 ] }) 1`] = `"1, 3-11"`;
